tests/EditorSuggest.test.ts

```diff
@@ -2,12 +2,10 @@
  * @jest-environment jsdom
  */
 import moment from 'moment';
-import { getSettings } from '../src/Config/Settings';
+import { getSettings, getTaskFormat } from '../src/Config/Settings';
```

Comment:
Picking a random place to add a comment. 😄 

I would find it really helpful if there were some specific, direct tests of the new functionality. I feel it would help reveal how the new code works, and would be helpful to anyone maintaining this code in future, or adding a new format.

I know that this code is tested indirectly by all the existing task reading & writing tests.

Perhaps there could be a new test which shows how to create a fictional, nonsense format that just supports one of each type of field... And shows reading and writing.

(Unless it's easier to support all of them.)

If that's not possible, maybe a few lines of tests in `DefaultTaskSerializer.test.ts` - even if it repeats a few lines of code in production code - which would be less useful for demo purposes, but would help for convenience of running simple tests when modifying

Reply:
Hi, thanks for asking.

I have very limited time in the next 2 days to reply, so please forgive a hasty reply.

I've evolved the tests quite a lot in the ~9 months I've been the maintainer, which does mean there is quite a lot of inconsistency in the code.

#### Task.test.ts

Generally I see the tests in `Task.test.ts` as the less useful tests we have.

#### Prefer custom matchers

I favour writing tests with Custom Matchers. They are harder to setup, but lead to much more expressive tests, and if their output on failure contains useful information about the context, there failures are much easier to diagnose.

https://github.com/obsidian-tasks-group/obsidian-tasks/commit/3d779ec113c3b6771a9cc97c103467bfed6f30c1

See:

- tests/CustomMatchers/CustomMatchersForFilters.ts and the other files in that directory
- And https://github.com/obsidian-tasks-group/obsidian-tasks/commit/3d779ec113c3b6771a9cc97c103467bfed6f30c1 which made them easier to use

#### Only test the local area

Tests for each layer of code should broadly assume that the lower layer has been well tested.

Now that the reading and writing code for Tasks has been moved out, probably all the pre-existing tests for that in Task.test.ts should be moved down to the 'default serialiser layer', and then maybe 'job mostly done'.

#### Demonstrating usage

Not to disagree with anything you said, and given the amount of thought you have already put in to it, I defer to you on how many tests you want to write...

But as an example, when someone wrote a test with powerful code that I didn't understand and would not have been able to maintain, they wrote a test for the core code, and it transformed the maintainability of the feature:

https://github.com/obsidian-tasks-group/obsidian-tasks/blob/main/tests/GroupingTreeNode.test.ts

That is very roughly the sort of thing that I would appreciate seeing for this new code. A simple demonstration - in executable code - of how it works/how to use it.

Reply:
I'm not sure if you are, but please don't agonise over perfect testing. Perfection can be the enemy of existing!

Reply:
The tests are great - very clear and definitely good enough. Thanks ever so much for that.